lib/scenic/themes.ex

```diff
@@ -0,0 +1,351 @@
+  """
+  @callback load() :: keyword
+  @optional_callbacks load: 0
+
+  defmacro __using__(sources \\ []) do
+    quote do
+      @behaviour Scenic.Themes
+      @opts_schema [
+        name: [required: true, type: :atom],
+        themes: [required: true, type: :any],
+        schema: [required: false, type: :any],
+        palette: [required: false, type: :any]
+      ]
+      @default_schema [:text, :background, :border, :active, :thumb, :focus]
+      @_palette %{}
+
+      @library_themes Enum.reduce(unquote(sources), %{}, fn lib_opts, acc ->
+        case NimbleOptions.validate(lib_opts, @opts_schema) do
+          {:ok, lib_opts} ->
+            name = lib_opts[:name]
+            themes = lib_opts[:themes]
+            schema = lib_opts[:schema] || []
+            palette = lib_opts[:palette] || %{}
+            @_palette Map.merge(@_palette, palette)
+            case themes do
+              themes when is_map(themes) ->
+                # not a module so we can load in the settings directly
+                Map.put_new(acc, name, {themes, List.flatten([@default_schema | schema])})
+              themes ->
+                # this is a module so we have to load the settings
+                lib_opts = themes.load()
+                themes = lib_opts[:themes]
+                schema = lib_opts[:schema] || []
+                palette = lib_opts[:palette] || %{}
+                @_palette Map.merge(@_palette, palette)
+                Map.put_new(acc, name, {themes, List.flatten([@default_schema | schema])})
+            end
+          {:error, error} ->
+            raise Exception.message(error)
+        end
+      end)
+
+      # validate the passed options
+      def library(), do: @library_themes
+
+      def _get_palette(), do: @_palette
+    end
+  end
+
+  @doc false
+  def module() do
+    with {:ok, config} <- Application.fetch_env(:scenic, :themes),
+         {:ok, module} <- Keyword.fetch(config, :module) do
+      module
+    else
+      _ ->
+        # No module configure return the default
+        __MODULE__
+    end
+  end
+
+  def validate(theme)
+  def validate({lib, theme_name} = lib_theme) when is_atom(theme_name) do
+    themes = module().library()
+    case Map.get(themes, lib) do
+      {themes, schema} ->
+        # validate against the schema
+        case Map.get(themes, theme_name) do
+          nil ->
+            {
+              :error,
+              """
+              #{IO.ANSI.red()}Invalid theme specification
+              Received: #{inspect(theme_name)}
+              #{IO.ANSI.yellow()}
+              The theme could not be found in library #{inspect(lib)}.
+              Ensure you got the name correct.
+              #{IO.ANSI.default_color()}
+              """
+            }
+          theme ->
+            case validate(theme, schema) do
+              {:ok, _} -> {:ok, lib_theme}
+              error -> error
+            end
+        end
+      nil ->
```

Comment:
I'm worried that there are many other projects that already use the standard themes of :primary, :secondary etc that will break as they need to be updated to {:scenic, :primary}. Instead of just nil -> it feels like if the theme_name is just an atom it should try {:scenic, :theme_name} here and see if it works. If that fails too, then go to the error.

Reply:
This would still allow the dev to override the meaning of :primary, since that would be found first.

Reply:
@boydm Personally deprecations are ok for me having in mind that scenic is below `1.0.0` version … Only after that I would expect change of first version number in order to introduce deprecation. I believe that every developer using `scenic` have in mind it's current state and it's not using it because this project have unbelievably stable API, but because it have amazing features other projects does not provide. :smiley:

Reply:
Went ahead and added backwards compat for single atom theme names for `normalize/1`, `preset/1`, `validate/1`

Reply:
It looks like `normalize/1` and `preset/1` are exactly the same. Maybe one of these can trigger validation and a raise and the other not?